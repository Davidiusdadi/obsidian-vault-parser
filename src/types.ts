export interface File {
  filePath: string;
  name: string;
  published: boolean;
  tags: string[];
  links: Set<string>;
  backLinks: Set<string>;
  contents: string;
  frontMatter: Record<string, any>;
}

export interface Vault {
  files: Record<string, File>;
}
