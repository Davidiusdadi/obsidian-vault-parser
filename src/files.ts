import path from "path";
import fs from "fs";

export const getFileName = (filepath: string): string =>
  path.parse(filepath).name;

export const readFile = async (path: string): Promise<string> =>
  fs.promises.readFile(path, "utf-8");

export const parseWikiLinks = (content: string): string[] => {
  const linkRegex = /\[\[([a-zA-Z0-9\s-]+\|?[a-zA-Z0-9\s]*)\]\]/g;
  const matches = Array.from(content.matchAll(linkRegex));
  return matches.map(m => {
    const splits = m[1].split("|");
    return splits[0];
  });
};
