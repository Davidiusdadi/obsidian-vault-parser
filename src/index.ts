import glob from "glob-promise";
import fs from "fs";
import path from "path";
import { File, Vault } from "./types";
import { getFileName } from "./files";

export const readFile = async (path: string): Promise<string> =>
  fs.promises.readFile(path, "utf-8");

export const parseFile = async (filepath: string): Promise<File> => {
  const contents = await readFile(filepath);
  const name = getFileName(filepath);

  return {
    filepath,
    name,
    links: [],
    tags: [],
    contents,
  };
};

export const emptyVault = (): Vault => ({
  files: {},
});

export const readVault = async (path: string): Promise<Vault> => {
  const files = await glob(`${path}/**/*.md`);

  const vault = emptyVault();

  for (const filepath of files) {
    const file = await parseFile(filepath);
    vault.files[file.name] = file;
  }

  return vault;
};
