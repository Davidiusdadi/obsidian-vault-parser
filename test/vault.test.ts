import path from "path";
import { getFileName, parseWikiLinks } from "../src/files";
import { readVault } from "../src/index";

describe("vault", () => {
  it("parse vault", async () => {
    const vaultPath = path.resolve(__dirname, "../test-vault");
    const vault = await readVault(vaultPath);

    expect(vault.files.hello.links).toEqual(["world", "projects"]);
    expect(vault.files.hello.backLinks).toEqual(["index", "world"]);

    expect(vault).toMatchSnapshot();
  });

  it("removes unpublished files from vault", async () => {
    const vaultPath = path.resolve(__dirname, "../test-vault");
    const vault = await readVault(vaultPath, {
      isPublished: f => !!f.frontMatter.published,
    });

    expect(vault.files.hello.links).toEqual(["projects"]);
    expect(vault.files.hello.backLinks).toEqual(["index"]);
  });
});

describe("files", () => {
  it("gets filename from filepath", () => {
    expect(getFileName("./hello.md")).toBe("hello");
    expect(getFileName("./foo/bar.md")).toBe("bar");
    expect(getFileName("/hello")).toBe("hello");
    expect(getFileName("/foo/bar")).toBe("bar");
    expect(getFileName("/foo/bar.md")).toBe("bar");
    expect(getFileName("/foo/hello world.md")).toBe("hello world");
    expect(getFileName("/foo/foo-bar.md")).toBe("foo-bar");
  });

  it("parses wikilinks", () => {
    expect(parseWikiLinks("hello")).toEqual([]);
    expect(parseWikiLinks("[[hello]]")).toEqual(["hello"]);
    expect(parseWikiLinks("hello [[world]]")).toEqual(["world"]);
    expect(parseWikiLinks("foo [[hello-world-123]] bar")).toEqual([
      "hello-world-123",
    ]);
    expect(parseWikiLinks("[[hello world]]")).toEqual(["hello world"]);

    expect(
      parseWikiLinks(`
Hello [[world]] this is a test

Woot [[woot]].
    `),
    ).toEqual(["world", "woot"]);

    // Aliases
    expect(parseWikiLinks(`[[hello|alias]]`)).toEqual(["hello"]);
    expect(parseWikiLinks(`[[hello|alias]] [[world]]`)).toEqual([
      "hello",
      "world",
    ]);
    expect(parseWikiLinks(`[[hello|another alias]]`)).toEqual(["hello"]);
  });
});
