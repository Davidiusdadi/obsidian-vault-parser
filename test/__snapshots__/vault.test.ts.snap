// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vault parse vault 1`] = `
Object {
  "files": Object {
    "hello": Object {
      "backLinks": Array [
        "index",
        "world",
      ],
      "content": "
This is the hello page!

go to [[world]]

hey there [[projects|project alias]]
",
      "frontMatter": Object {
        "published": true,
      },
      "links": Array [
        "world",
        "projects",
      ],
      "name": "hello",
      "path": "/Users/jakerunzer/dev/obsidian-vault-parser/test-vault/hello.md",
      "tags": Array [],
    },
    "index": Object {
      "backLinks": Array [],
      "content": "
# Hello world

[[hello]]",
      "frontMatter": Object {
        "published": true,
      },
      "links": Array [
        "hello",
      ],
      "name": "index",
      "path": "/Users/jakerunzer/dev/obsidian-vault-parser/test-vault/index.md",
      "tags": Array [],
    },
    "my-awesome-project": Object {
      "backLinks": Array [
        "projects",
      ],
      "content": "
# My Project

Hello, this is my awesome project
",
      "frontMatter": Object {
        "published": true,
      },
      "links": Array [],
      "name": "my-awesome-project",
      "path": "/Users/jakerunzer/dev/obsidian-vault-parser/test-vault/projects/my-awesome-project.md",
      "tags": Array [],
    },
    "projects": Object {
      "backLinks": Array [
        "hello",
      ],
      "content": "
Some projects are

- [[my-awesome-project]]
",
      "frontMatter": Object {
        "published": true,
      },
      "links": Array [
        "my-awesome-project",
      ],
      "name": "projects",
      "path": "/Users/jakerunzer/dev/obsidian-vault-parser/test-vault/projects.md",
      "tags": Array [],
    },
    "world": Object {
      "backLinks": Array [
        "hello",
      ],
      "content": "[[hello]] world!
",
      "frontMatter": Object {},
      "links": Array [
        "hello",
      ],
      "name": "world",
      "path": "/Users/jakerunzer/dev/obsidian-vault-parser/test-vault/world.md",
      "tags": Array [],
    },
  },
  "path": "/Users/jakerunzer/dev/obsidian-vault-parser/test-vault",
}
`;
